pkg/controller/core/workload_controller.go

```diff
@@ -166,9 +160,51 @@ func (r *WorkloadReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 		return ctrl.Result{}, nil
 	}
 
+	if ptr.Deref(wl.Spec.Active, true) {
+		var updated bool
+
+		// If a deactivated workload is re-activated we need to reset the RequeueState.
+		if workload.IsEvictedByDeactivation(&wl) && wl.Status.RequeueState != nil {
+			wl.Status.RequeueState = nil
+			updated = true
+		}
+
+		if cond := apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadRequeued); cond != nil && cond.Status == metav1.ConditionFalse {
+			switch cond.Reason {
+			case kueue.WorkloadEvictedByDeactivation:
+				workload.SetRequeuedCondition(&wl, kueue.WorkloadReactivated, "The workload was reactivated", true)
+				updated = true
+			case kueue.WorkloadEvictedByPodsReadyTimeout:
+				var requeueAfter time.Duration
+				if wl.Status.RequeueState != nil && wl.Status.RequeueState.RequeueAt != nil {
+					requeueAfter = wl.Status.RequeueState.RequeueAt.Time.Sub(r.clock.Now())
+				}
+				if requeueAfter > 0 {
+					return reconcile.Result{RequeueAfter: requeueAfter}, nil
+				}
+				workload.SetRequeuedCondition(&wl, kueue.WorkloadBackoffFinished, "The workload backoff was finished", true)
```

Comment:
It looks inconsistent: that we don't clear`wl.Status.RequeueState` for PodsReady timeout, but we clear it on L168 in case of Reactivation. In particular, I'm wondering if this could lead to incrementing the `RequeueState.Count` [here](https://github.com/kubernetes-sigs/kueue/blob/f108fef22eb4532f3918754e5a08226ac366343c/pkg/controller/core/workload_controller.go#L420) on the second requeue. Please verify and fix if this is the case.

Reply:
I'm not sure that we need to clear `wl.Status.RequeueState` in this case, because we need to know how much attempts we have, to deactivate workload if we exceeded the limit [here](https://github.com/kubernetes-sigs/kueue/blob/f108fef22eb4532f3918754e5a08226ac366343c/pkg/controller/core/workload_controller.go#L421).

```go
func (r *WorkloadReconciler) triggerDeactivationOrBackoffRequeue(ctx context.Context, wl *kueue.Workload) (bool, error) {
...
        // If requeuingBackoffLimitCount equals to null, the workloads is repeatedly and endless re-queued.
	requeuingCount := ptr.Deref(wl.Status.RequeueState.Count, 0) + 1
	if r.waitForPodsReady.requeuingBackoffLimitCount != nil && requeuingCount > *r.waitForPodsReady.requeuingBackoffLimitCount {
		wl.Spec.Active = ptr.To(false)
		if err := r.client.Update(ctx, wl); err != nil {
			return false, err
		}
		r.recorder.Eventf(wl, corev1.EventTypeNormal, kueue.WorkloadEvictedByDeactivation,
			"Deactivated Workload %q by reached re-queue backoffLimitCount", klog.KObj(wl))
		return true, nil
	}
...
}
```

Reply:
Or you mean to clear just RequeueAt field?

Reply:
> I'm not sure that we need to clear wl.Status.RequeueState in this case, because we need to know how much attempts we have

Ah, got it, we need to preserve the `count` so that we can increment it if the PodsReady timeout is exceeded for the next time.

> Or you mean to clear just RequeueAt field?

I guess the field is harmless, but might be confusing, since the workload is already requeued, so indeed, I would suggest to clear it.